lib/LANraragi/Plugin/Metadata/HentagOnline.pm

```diff
@@ -0,0 +1,238 @@
+    }
+
+    my $vault_id;
+    my @source_urls = undef;
+
+    # First, try running based on a vault ID from a hentag URL
+    if (defined($oneshot_param) && $oneshot_param ne '') {
+        $logger->info("Oneshot param = $oneshot_param");
+        # URL was passed as a oneshot param, done!
+        $vault_id = parse_vault_url($oneshot_param);
+        if (!defined($vault_id)) {
+            $logger->info("Failed parsing URL $oneshot_param");
+            # Don't continue execution if the oneshot URL was invalid. Raise an error instead to avoid surprises
+            return ( error => "Did not recognize the URL, is it a proper vault URL?" );
+        }
+    } elsif (my $hentag_source = get_existing_hentag_source_url(@existing_tags)) {
+        # There might be a hentag URL inside a tag. If it fails parsing, that's okay, just continue execution with any other approach
+        if (defined($hentag_source)) {
+            $logger->info("Found an existing (parsable) hentag source, will use that!");
+            $vault_id = parse_vault_url($hentag_source);
+        }
+    } else {
+        # Try to fetch any other existing source tags
+        @source_urls = get_source_tags(@existing_tags);
+    }
+
+    my $stringjson = '';
+    if (defined($vault_id)) {
+        # Use a vault ID for lookups
+        $logger->info('Vault ID run');
+        $stringjson = get_json_by_vault_id($ua, $vault_id, $logger);
+    } elsif (@source_urls) {
+        # Use existing URL tags for lookups
+        $logger->info('URL lookup run');
+        $stringjson = get_json_by_urls($ua, $logger, @source_urls);
+    } else {
+        # Title lookup
+        $logger->info('Title lookup');
+        my $archive_title = $lrr_info->{archive_title};
+
+        if ($stringjson eq '') {
+            $stringjson = get_json_by_title($ua, $archive_title, $logger);
+        }
+    }
+
+    if ($stringjson ne '') {
+        $logger->debug("Received the following JSON: $stringjson");
+        my $json = from_json($stringjson);
+
+        #Parse it
+        my ( $tags, $title ) = tags_from_hentag_api_json($json, $allowed_languages);
+
+        #Return tags IFF data is found
+        $logger->info("Sending the following tags to LRR: $tags");
+        if ( $save_title && $title ) {
+            $logger->info("Parsed title is $title");
+            return ( tags => $tags, title => $title );
+        } elsif ($tags ne "") {
+            return ( tags => $tags );
+        }
+    }
+
+    return ( error => "No matching Hentag Archive Found!" );
+}
+
+# Returns the ID from a hentag URL, or undef if invalid.
+sub parse_vault_url($url) {
+    if (!defined $url) {
+        return;
+    }
+    if ( $url =~ /https?:\/\/(?:www\.)?hentag\.com\/vault\/([a-zA-Z0-9]*)\/?.*/ ) {
+        return $1;
+    }
+    return;
+}
+
+# Fairly good for mocking in tests
+# get_json_by_title(ua, archive_title, logger)
+sub get_json_by_title($ua, $archive_title, $logger) {
+    my $stringjson = '';
```

Comment:
should probably be `string_json` to match the other subs

Reply:
Fixed!